src/Metadata/CIP25.purs

```diff
@@ -0,0 +1,261 @@
+
+--------------------------------------------------------------------------------
+-- CIP25Metadata
+--------------------------------------------------------------------------------
+
+newtype CIP25Metadata = CIP25Metadata (Array CIP25MetadataEntry)
+
+derive instance Generic CIP25Metadata _
+derive instance Newtype CIP25Metadata _
+derive instance Eq CIP25Metadata
+
+instance Show CIP25Metadata where
+  show = genericShow
+
+instance ToData CIP25Metadata where
+  toData (CIP25Metadata entries) = unsafePartial $ toData
+    $ Map.singleton (unsafeMkKey nftMetadataLabel)
+    $ Map.fromFoldable
+    $ flip map groups
+    $ \group ->
+        toData (_.policyId $ NonEmpty.head group) /\
+          (Map.fromFoldable <<< flip map group) \entry ->
+            toData (entry.assetName) /\ metadataEntryToData_ (wrap entry)
+    where
+    groups :: Array (NonEmptyArray CIP25MetadataEntryRecord)
+    groups = flip groupBy (unwrap <$> entries) $
+      \a b -> a.policyId == b.policyId
+
+instance FromData CIP25Metadata where
+  fromData meta = unsafePartial $ do
+    entries <- lookupKey nftMetadataLabel meta >>= case _ of
+      Map mp1 -> map concat
+        $ for mp1
+        $ \(policyId /\ assets) ->
+            case assets of
+              Map mp2 ->
+                for mp2 $ \(assetName /\ contents) ->
+                  metadataEntryFromData_ <$> fromData policyId
+                    <*> fromData assetName
+                    <*> pure contents
+              _ -> Nothing
+      _ -> Nothing
+    wrap <$> sequence entries
+
+instance DecodeJson CIP25Metadata where
+  decodeJson =
+    Json.caseJsonObject errExpectedObject $ \obj -> do
+      policies <- obj .: nftMetadataLabel
+      caseJsonObject policies $ \objPolicies ->
+        map (wrap <<< concat)
+          $ for (objToArray objPolicies)
+          $ \(policyId /\ assets) ->
+              caseJsonObject assets $ \objAssets ->
+                for (objToArray objAssets) $ \(assetName /\ contents) -> do
+                  policyId_ <- decodePolicyId policyId
+                  assetName_ <- decodeAssetName assetName
+                  metadataEntryDecodeJson_ policyId_ assetName_ contents
+    where
+    objToArray :: forall a. FO.Object a -> Array (Tuple String a)
+    objToArray = FO.toUnfoldable
+
+    caseJsonObject
+      :: forall a
+       . Json
+      -> (FO.Object Json -> Either Json.JsonDecodeError a)
+      -> Either Json.JsonDecodeError a
+    caseJsonObject = flip (Json.caseJsonObject errExpectedObject)
+
+    decodePolicyId :: String -> Either Json.JsonDecodeError MintingPolicyHash
+    decodePolicyId =
+      note (Json.TypeMismatch "Expected hex-encoded policy id")
+        <<< map wrap
+        <<< (scriptHashFromBytes <=< hexToByteArray)
+
+    decodeAssetName :: String -> Either Json.JsonDecodeError TokenName
+    decodeAssetName =
+      note (Json.TypeMismatch "Expected UTF-8 encoded asset name")
+        <<< (mkTokenName <=< byteArrayFromString)
```

Comment:
I'm not exactly sure how we should handle this, but the `<asset_name>` key is indeed UTF-8 (for now) according to the CIP. But `byteArrayFromString` is only for ASCII strings -- so if anything code points in the asset name fall outside of the ASCII range we'll get a `Nothing` back for the whole thing

Reply:
I guess we should use `TextEncoder` and `TextDecoder` JS APIs here. And now we should rename `byteArrayFromString` to something like ` byteArrayFromASCII`.

In Seabug metadata we didn't have arbitrary utf-8 strings, which is the reason why this problem didn't hit us earlier.